scripts/bert_SQuAD/dataset.py

```diff
@@ -0,0 +1,493 @@
+                 end_position,
+                 is_impossible):
+        self.example_id = example_id
+        self.qas_id = qas_id
+        self.doc_tokens = doc_tokens
+        self.doc_span_index = doc_span_index
+        self.tokens = tokens
+        self.token_to_orig_map = token_to_orig_map
+        self.token_is_max_context = token_is_max_context
+        self.input_ids = input_ids
+        self.valid_length = valid_length
+        self.segment_ids = segment_ids
+        self.start_position = start_position
+        self.end_position = end_position
+        self.is_impossible = is_impossible
+
+
+class SQuAD(SimpleDataset):
+    """Stanford Question Answering Dataset (SQuAD) - reading comprehension dataset.
+
+    From
+    https://rajpurkar.github.io/SQuAD-explorer/
+
+    License: CreativeCommons BY-SA 4.0
+
+    The original data format is json, which has multiple contexts (a context is a paragraph of text
+    from which questions are drawn). For each context there are multiple questions, and for each of
+    these questions there are multiple (usually 3) answers.
+
+    This class loads the json and flattens it to question dataset.
+    Number of records in the dataset is equal to number of questions in json file.
+
+
+    Parameters
+    ----------
+    filename : str
+        SQuAD json file path.
+    is_training : bool, default True
+        Whether to run training.
+    version_2 : bool, default False
+        If true, the SQuAD examples contain some that do not have an answer.
+    """
+
+    def __init__(self, filename, is_training=True, version_2=False):
+        self.input_file = filename
+        self.is_training = is_training
+        self.version_2 = version_2
+        super(SQuAD, self).__init__(self._read())
+
+    def _read(self):
+        """Read a SQuAD json file into a list of SquadExample."""
+        with open(self.input_file, 'r') as reader:
+            input_data = json.load(reader)['data']
+
+        def is_whitespace(c):
+            if c == ' ' or c == '\t' or c == '\r' or c == '\n' or ord(
+                    c) == 0x202F:
+                return True
+            return False
+
+        examples = []
+        example_id = 0
+        for entry in input_data:
+            for paragraph in entry['paragraphs']:
+                paragraph_text = paragraph['context']
+                doc_tokens = []
+                char_to_word_offset = []
+                prev_is_whitespace = True
+                for c in paragraph_text:
+                    if is_whitespace(c):
+                        prev_is_whitespace = True
+                    else:
+                        if prev_is_whitespace:
+                            doc_tokens.append(c)
+                        else:
+                            doc_tokens[-1] += c
+                        prev_is_whitespace = False
+                    char_to_word_offset.append(len(doc_tokens) - 1)
+
+                for qa in paragraph['qas']:
```

Comment:
It looks these can also be built as "Transforms" on top of the SQuAD dataset. https://github.com/dmlc/gluon-nlp/blob/master/src/gluonnlp/data/question_answering.py#L36 What is the reason the SQuAD dataset API is not reused?